File: backend/src/main/java/com/kiwi/features/skills/data/SkillDomain.java
package com.kiwi.features.skills.data;

import lombok.*;
import java.time.Instant;

@Getter
@Setter
public class SkillDomain {

    private Long skillId;
    private String name;
    private String description;
    private String quote;
    private SkillType type;

    private CooldownType cooldownType;
    private Long cooldownGoalId;
    private Integer cooldownTimeMinutes;
    private String cooldownOtherDescription;
    private Long levelupSkillId;

    private boolean isCooldown;
    private Instant cooldownUntil;
    private int deckSlot;

    public SkillDomain(
            Long skillId,
            String name,
            String description,
            String quote,
            SkillType type,
            CooldownType cooldownType,
            Long cooldownGoalId,
            Integer cooldownTimeMinutes,
            String cooldownOtherDescription,
            Long levelupSkillId,
            boolean isCooldown,
            Instant cooldownUntil,
            int deckSlot
    ) {
        this.skillId = skillId;
        this.name = name;
        this.description = description;
        this.quote = quote;
        this.type = type;
        this.cooldownType = cooldownType;
        this.cooldownGoalId = cooldownGoalId;
        this.cooldownTimeMinutes = cooldownTimeMinutes;
        this.cooldownOtherDescription = cooldownOtherDescription;
        this.levelupSkillId = levelupSkillId;
        this.isCooldown = isCooldown;
        this.cooldownUntil = cooldownUntil;
        this.deckSlot = deckSlot;
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/features/skills/data/SkillDomain.kt
package com.bellako.kiwi.features.skills.data

import java.time.Instant

sealed class SkillDomain {
    abstract val id: Long
    abstract val name: String
    abstract val type: SkillType
    abstract val description: String
    abstract val quote: String?
    abstract val icon: Int
    abstract val onThrowSFX: Int
    abstract val levelupSkillId: Long?
    abstract val isCooldown: Boolean
    abstract val deckSlot: Int

    data class Other(
        override val id: Long,
        override val name: String,
        override val type: SkillType,
        override val description: String,
        override val quote: String?,
        override val icon: Int,
        override val onThrowSFX: Int,
        override val levelupSkillId: Long?,
        override val isCooldown: Boolean,
        override val deckSlot: Int,
        val cooldownOtherDescription: String,
    ) : SkillDomain()

    data class Time(
        override val id: Long,
        override val name: String,
        override val type: SkillType,
        override val description: String,
        override val quote: String?,
        override val icon: Int,
        override val onThrowSFX: Int,
        override val levelupSkillId: Long?,
        override val isCooldown: Boolean,
        override val deckSlot: Int,
        val cooldownTimeMinutes: Int,
        val cooldownUntil: Instant?,
        val cooldownProgress: Float,
    ) : SkillDomain()

    data class Goal(
        override val id: Long,
        override val name: String,
        override val type: SkillType,
        override val description: String,
        override val quote: String?,
        override val icon: Int,
        override val onThrowSFX: Int,
        override val levelupSkillId: Long?,
        override val isCooldown: Boolean,
        override val deckSlot: Int,
        val cooldownGoalId: Long,
        val goalData: GoalData,
    ) : SkillDomain()
}

data class GoalData(
    val action: String,
    val progress: Int,
    val target: Int,
)
